### DISCUSSION TITLE
Don't select unnecessary columns

### DISCUSSION THREAD

**Discussion by artelk** [linq2db/linq2db#3168] [Q&A] [answered]
```cs
DataConnection.TurnTraceSwitchOn();
DataConnection.WriteTraceLine = ...; // write SQL request

bool b = false;
var q = from t in db.Table
             select new
             {
                Column1,
                Other = b ? new { t.Column2, t.Column3 } : null
             };
q.ToList();
```
The SQL generated selects all 3 columns even if the `b` was false.
What can I do to not load Column2 and Column3 in that case?

**Comment by MaceWindu** [comment]
Not sure it will work, but you can try to force linq2db to evaluate `Other` on server:
```cs
Other = Sql.AsSql(b ? new { t.Column2, t.Column3 } : null)
```

**Reply by artelk** [reply]
I wanted something opposite. With `false` hardcoded it doesn't select Column2 and Column3. But with the local variable `b` it selects those columns even if `b` is `false` when you execute the query. (Btw, can I trust `DataConnection.WriteTraceLine` if it really catches the SQL when it is being sent to the server?)

**Reply by MaceWindu** [reply]
yes, trace contains real sql (except parameters section, generated by logger): https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/Data/TraceInfo.cs#L128

**Reply by artelk** [reply]
Ok, so it really loads unnecessary columns. How can I avoid that?

**Comment by artelk** [comment, marked as answer]
Found workaround
```cs
[ExpressionMethod(nameof(MyIIFImpl))]
public static T MyIIF<T>(bool b, T x, T y) => b ? x : y;

private static Expression<Func<bool, T, T, T>> MyIIFImpl<T>() =>
  (b, x, y) => MyIIFImpl<T>(b).Compile()(x, y);

private static Expression<Func<T, T, T>> MyIIFImpl<T>(bool b) =>
  b ? (x, y) => x : (x, y) => y;

bool b = false;
var q = from t in db.Table
        select new
        {
           Column1,
           Other = MyIIF(b, new { t.Column2, t.Column3 } , null)
        };
```
The MyIIF requires the parameter `b` to be calculated on client side before executing the query.

But it is better to make this default behavior, no reason to load columns if they aren't used at all when the condition can be calculated on client side and it is known what branch of `?:` operator to chose.
